pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/BookkeeperBucketSnapshotStorage.java

```diff
@@ -0,0 +1,246 @@
+import org.apache.pulsar.broker.delayed.proto.DelayedMessageIndexBucketSnapshotFormat.SnapshotMetadata;
+import org.apache.pulsar.broker.delayed.proto.DelayedMessageIndexBucketSnapshotFormat.SnapshotSegment;
+import org.apache.pulsar.common.util.FutureUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BookkeeperBucketSnapshotStorage implements BucketSnapshotStorage {
+
+    private static final Logger log = LoggerFactory.getLogger(BookkeeperBucketSnapshotStorage.class);
+    private static final byte[] LedgerPassword = "".getBytes();
+
+    private final PulsarService pulsar;
+    private final ServiceConfiguration config;
+    private BookKeeper bookKeeper;
+
+    public BookkeeperBucketSnapshotStorage(PulsarService pulsar) {
+        this.pulsar = pulsar;
+        this.config = pulsar.getConfig();
+    }
+
+    @Override
+    public CompletableFuture<Long> createBucketSnapshot(SnapshotMetadata snapshotMetadata,
+                                                        List<SnapshotSegment> bucketSnapshotSegments) {
+        return createLedger()
+                .thenCompose(ledgerHandle -> addEntry(ledgerHandle, snapshotMetadata.toByteArray())
+                                .thenApply(__ -> ledgerHandle))
+                .thenCompose(ledgerHandle -> addSnapshotSegments(ledgerHandle, bucketSnapshotSegments)
+                        .thenApply(__ -> ledgerHandle))
+                .thenCompose(ledgerHandle -> closeLedger(ledgerHandle)
+                        .thenApply(__ -> ledgerHandle.getId()));
+    }
+
+    @Override
+    public CompletableFuture<SnapshotMetadata> getBucketSnapshotMetadata(long bucketId) {
+        return openLedger(bucketId).thenCompose(
+                ledgerHandle -> getLedgerEntry(ledgerHandle, 0, 0).
+                        thenCompose(
+                        entryEnumeration -> parseSnapshotMetadataEntry(entryEnumeration.nextElement())));
+    }
+
+    @Override
+    public CompletableFuture<List<SnapshotSegment>> getBucketSnapshotSegment(long bucketId, long firstSegmentEntryId,
+                                                                             long lastSegmentEntryId) {
+        return openLedger(bucketId).thenCompose(
+                ledgerHandle -> getLedgerEntry(ledgerHandle, firstSegmentEntryId, lastSegmentEntryId).thenCompose(
+                        this::parseSnapshotSegmentEntries));
+    }
+
+    @Override
+    public CompletableFuture<Long> getBucketSnapshotLength(long bucketId) {
+        return openLedger(bucketId).thenApply(ledgerHandle -> {
+            long length = ledgerHandle.getLength();
+            closeLedger(ledgerHandle);
+            return length;
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> deleteBucketSnapshot(long bucketId) {
+        return deleteLedger(bucketId);
+    }
+
+    @Override
+    public void start() throws Exception {
+        this.bookKeeper = pulsar.getBookKeeperClientFactory().create(
```

Comment:
Do we really need a separate BK client?

Reply:
The`defaultBkClient` uses `statsLogger` to record operate stat of `managedLedger`, I don't think should use the `defaultBkClient`.
https://github.com/apache/pulsar/blob/6fec66b12b04a37e4c2b05d78d4e33b380c270df/pulsar-broker/src/main/java/org/apache/pulsar/broker/ManagedLedgerClientFactory.java#L84-L89

Reply:
With this newly declared bk-client, there are now three bk-clients in pulsar, and I still think it is too expensive to use a single bk-client for the delay message bucket. Maybe a separate discussion should be held later to resolve this issue.

Reply:
BK client creates other threads, see: 

https://github.com/apache/bookkeeper/blob/a3401a2139a554a8fbe52c8a398eeb4e51a8dad1/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeper.java#L426-L435

Maybe we can not fix the problem in this PR for now, and we can solve it all( `MangedLedger`, `Schema`,`Delay message` ) together by a new DISCUSS.

Reply:
>I still think it is too expensive to use a single bk-client for the delay message

Hi @poorbarcode 
Could you indicate which part is expensive? then we can complete the trade-off.

Reply:
> If you want to use bookkeeper storage, you need a client. Maybe if we have more service needs for an internal bookie client, we can create only one internal bookie client and then share it with different components.

+1

Hi @mattisonchao 

> Could you indicate which part is expensive? then we can complete the trade-off.

A separate bookie client for the delay message bucket only.  see also https://github.com/apache/pulsar/pull/17677#discussion_r1025944991